modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+%dw 2.0
+output application/json
+
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a closed object is `{| |}`. The following example makes the `User` type support only the keys `firstName` and `lastName`. The script throws an exception because the `age` field is not accepted.
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Function Type
+
+DataWeave is a functional language, and functions are first class citizens, which means that functions are values,
+and values have an associated type.
+
+The syntax to define a function type is:
+
+----
+(paramType: ParamType,...) -> ReturnType
+----
+
+For example, if you want to define in a constraint a `Function` type that has a parameter of type `String`, a second parameter of type `Number`, and a `Boolean` return value, the correct syntax is:
+
+----
+(paramA: String, paramB: Number) -> Boolean
+----
+
+For example, you can define a function that receives another function as an argument:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???
+----
+
+If you call the function `applyIDsChange` with a function that does not match the `changeTo` constraint `(Number) -> Number`, DataWeave throws a type-checking error.
+For example, the following call works because `abs` is a function that takes a value of the `Number` type and returns a `Number`, which matches the `changeTo` parameter.
+
+----
+applyIDsChange([1,-6, 3, -8], abs)
+----
+
+But this call fails because `sum` is a function that takes an Array and returns a Number, so it does not match our parameter constraint.
+
+----
+applyIDsChange([1,-6, 3, -8], sum)
+----
+
+== Complex Types
+
+=== Any and Nothing
+
+In some cases, you cannot enforce a restriction because the type accepts all the values. +
+
+* `Any` type accepts all possible values.
+* `Nothing` type accepts no value, but it can be assigned to all the types. This type is not frequently used explicitly, but it is used by the type inference algorithm.
+
+=== Union Type
+
+The `Union` type is used to compose types. The syntax to define a `Union` type is:
+
+----
+TypeA | TypeB | ...
+----
+
+The following example defines a variable with a constraint that accepts a value of type `String` or `Number`:
+
+----
+var age: String | Number = if (payload.allStrings) "32" else 32
+----
+
+A common pattern is to use `Null` type with a `Union` type to specify that a type accepts `null` as a value. In the following example, the variable `a` allows the expression `payload.foo` to return `null`.
+
+----
+var a: String| Null = payload.foo
+----
+
+
+=== Intersection Type
+
+The `Intersection` type intersects `Object` types. In this case, the intersection works as the concatenation (`++`) of object types.
```

Comment:
Please define what "intersects" means in this context: creates an overlap?

Also...avoid using italic inline headings like the one following. Merge the version info into the text:
```suggestion
"Introduced in DataWeave 2.3.0 and supported by Mule 4.3 and later, the `Intersection` type intersects `Object` types. In this case, the intersection works as the concatenation (`++`) of object types.
```